### ISSUE TITLE
Some css comments are not being recognized

### ISSUE THREAD

**Issue by personaaleatoria** [csstree/csstree#142]
When parsing the bootstrap css file https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.5.0/css/bootstrap.min.css

I am able to see some comments in the ast but not the last one

`/*# sourceMappingURL=bootstrap.min.css.map */`

since this file shows a 404 error when not using the nodejs version I wanted to remove it using css-tree but it seems unable to parse it and add it to the ast.

You can also try with this simple example

```
/*!
 * this is a comment.
 */

.title {
  color: green;
}

/*# sourceMappingURL=bootstrap.min.css.map */
```

**Comment by lahmatiy** [comment]
CSSTree ignores comments expect for exclamation comments (i.e. `/*! .. */`) on top level. We may add top level `/*# .. */` comments as well. However, I'm going to remove comments from AST completely (with whitespaces as well). My suggestion is to use the [onComment](https://github.com/csstree/csstree/blob/master/docs/parsing.md#oncomment) option of parser to get all the comments and handle them somehow.

**Comment by kafwangure** [comment]
I'm not yet familiar with the CSSTree project. Is there a technical reason for omitting comments? I ask this because pretty much every parser I've worked with tries  to maintain the following invariant:

```javascript
stringify(parse(content)) === content
```
Also, is there a canonical/recommended way of "injecting back" the comment into the AST? I would have imagined an API like the following would have been possible based on other parsers/ast-walkers I've worked with.

```javascript
parse(content, {
     onComment(_value, _loc) {
       this.consume(Comment)
    }
})
```

**Comment by milkbump** [comment]
> Is there a technical reason for omitting comments?

👀

**Comment by lazarljubenovic** [comment]
None of the parser I've encountered are strict with `stringify(parse(content)) === content` in terms of trivia (whitespace, comments). Of course, it's not like I'm some parser guru, so that doesn't mean I'm speaking for literally every parser under the sun, so that with a grain of salt. Usually there's a switch that controls if trivia is saved or not, depending on your needs.

**Comment by osteele** [comment]
I had expected csstree to preserve comments, based on the description “focus on […] source-to-source transforming tasks” at the top of the README.

As an (appreciative) user of the library, IMO it would be reasonable to declare preserving comments out of scope for the project; but in that case it would be helpful to document this limitation in an equally prominent location. For example, by adding a bullet item to the Features list, e.g. "**Source-to-source transformation**. CSSTree can be used to modify a CSS document, by calling `parse`, `walk`, and `generate` in succession. Note that this is not true source-to-source transformation, as it discards comments that don't start with exclamation marks."

Again, this is not to dis the project or its creators and maintainers. I just think this would help developers to discover whether the library is appropriate for their use cases earlier in the evaluation process.

**Comment by lahmatiy** [comment]
I agree that losing formatting and comments can be confusing. Adding a note to the README about the limitations of source-to-source transformation is a good suggestion, and I believe it's worth including.

By the way, I'm considering alternative solutions for CSS transformations that would meet expectations while preserving both formatting and comments. I think the solution should be based on "definition syntax" and tokens rather than the AST. However, this approach requires research and development time that I currently don't have. So for now, adding a note about the limitation is the best option.

**Comment by tonyganchev** [comment]
I'm the maintainer of _@tony.ganchev/eslint-plugin-header_. It is validating the structure of standard copyright headers in sources and for that it relies on the configured parsers to pass the necessary comment objects. Given that a lot of ESlint-related parsers for CSS, especially the now default _@eslint/css_ one work on top of CSS tree, to support linting CSS I need to have a completely different approach. In other words - the README noticed helped me realize where the issue was, but that is it.

I am also looking forward to understanding the rationale behind this decision.
